File: cpp/euler4.cpp
#include <iostream>

template <typename T>
bool is_palindrome(T candidate)
{
	T n = candidate, rev = 0;
	while (n > 0)
	{
		rev = rev * 10 + n % 10;
		n /= 10;
	}
	return candidate == rev;
}

int main(int argc, char *argv[])
{
	int largest_palindrome = 0, cx, cy;
	int max = 1000, min = 100;
	for (int x = min; x < max; ++x)
	{
		for (int y = min; y < max; ++y)
		{
			if (is_palindrome(x * y))
			{
				int product = x * y;
				if (product > largest_palindrome)
				{
					cx = x; cy = y;
					largest_palindrome = product;
				}
			}
		}
	}
	std::cout << "Answer is "
			  << largest_palindrome
			  << " (" << cx << " * " << cy << ")"
			  << std::endl;
	return 0;
}

File: lua/euler4.lua
#!/usr/bin/env lua

function is_palindrome(candidate)
	local n, reversed = candidate, 0
	while n > 0 do
		reversed = reversed * 10 + n % 10
		n = math.floor(n / 10)
	end
	return candidate == reversed
end

function euler4()
	local min, max, largest = 100, 1000, 0
	for x = min, max do
		for y = min, max do
			if is_palindrome(x * y) then
				local product = x * y
				if product > largest then
					largest = product
				end
			end
		end
	end

	return largest
end

print("Answer is " .. euler4())
